update security best practices section

lirantal opened the pull request:
Updating existing and new sections.
Addresses #33  and #133

BrunoScheufler commented:
@lirantal awesome! Will have a look at it later 👍

i0natan commented:
@lirantal looks great, have few questions/comments, will share by tomorrow

i0natan commented:
@BrunoScheufler doesn' really matter, maybe leave open so we discuss my questions here (will share today)

BrunoScheufler commented:
Okay, I'll review it today :+1:

i0natan commented:
@lirantal I think it's great stuff, great advice, and we're pretty close to 20+.

I'm putting aside minor ideas I have in mind (e.g. CI instead of Git hook), we can do some minor refinements phase in the end. I'm now focused on the core question - is the current TOC comprehensive and attractive enough? 

I believe we can improve 2 things:
1. More Node related - the main 'threat' IMO, is that it will strike to the reader as generic security things ('validate input', 'beware of Regx' etc). What makes the difference is a list of very practical and Node-related things so I suggest including Node narratives in every title and make the titles interesting. For example, 

Secure Dependency Management - can be: Automatically detect NPM/Yarn vulnerabilities
Implement revocation of API tokens - can be: Allow token revocation with Passport & black list
Store passwords in a secure manner - can be: Encrypt passwords using the Bcrypt module (less generic and accurate, more practical and Node-relateted)

2. The prioritization/ordering might need some updates. For example, revocation of tokens is probably less important than detecting vulenerabilities but ranked much higer

Does this make sense?

BrunoScheufler commented:
@i0natan
> Secure Dependency Management - can be: Automatically detect NPM/Yarn vulnerabilities
I think we have added this one multiple times (4.4 5.13), it would at least be duplicated in the security best practices secion.

Other than this, how should we continue?

BrunoScheufler commented:
@i0natan that sounds good, we should have it in the security best practices section only anyways.
I think prioritizing content is a good idea, we should discuss this.

lirantal commented:
@i0natan addressing your comments:

> More Node related - the main 'threat' IMO, is that it will strike to the reader as generic security things ('validate input', 'beware of Regx' etc). What makes the difference is a list of very practical and Node-related things so I suggest including Node narratives in every title and make the titles interesting.

* Regular Expressions DoS is actually a very big concern for JavaScript due to the single-thread nature of the platform. To give an example, just a few months ago the popular `moment` library suffered from a regex dos vulnerability. We can use these real-world examples to convey the important of understanding these security concerns since probably everyone knows moment.
* Input validation is generic but important too, we can name it better probably to have this catching narrative as you suggested.

> The prioritization/ordering might need some updates. For example, revocation of tokens is probably less important than detecting vulenerabilities but ranked much higer

Yep, I didn't really change or take action on ordering and we should probably update it.

lirantal commented:
@BrunoScheufler agree, the secure dependencies management should be removed from the production check-list or other sections and exist in the security part.

i0natan commented:
> JavaScript due to the single-thread nature of the platform

This is a great example. I always knew that RegEx can be evil but never thought it's somehow more relevant in Node apps. When I first read it in the list, I felt like 'another RegEx tip'. Now when you wrote that explanation it striked me how important this is in the Node platform. For that reason, we must put the node aspect in the title:

Example, instead of:
Beware of RegEx

Make it Node related:
Prevent malicious RegEx from overloading your single-thread (or any other text which provides the Node context)

The example you brought is anotehr winning differentiator - we must include these examples

lirantal commented:
I'm just happy we're having a great discussion on this - totally raises awareness for readers.
I like the title you phrased and will update the docs with it + the example.

BrunoScheufler commented:
@lirantal it really is working out like a charm!
